### ISSUE TITLE
Problem using crawler

### ISSUE THREAD

**Issue by Pelirrojo** [Machine-Learning-Labs/vgchartzScrape#1]
From another [issue](https://github.com/GregorUT/vgchartzScrape/issues/6), thanks @JulesMillen 👍🏻  

> Hi Pelirrojo,
> 
> This is the only way I could think to contact you - so I hope this reaches you...
> 
> I have just downloaded and tried to run your version of the vgchartzScrape.
> 
> It seems to run okay but it is only outputting 200 rows of data, most of which are missing information from most columns (this issue is occurring on multiple computers). I am not sure whether this is an issue at my end (I dont think it is, however I am extremely novice with python) or whether something has changed in the website and so the code needs to be updated?
> 
> Any help would be hugely appreciated!
> 
> Alternatively, if you happen to have a scrape of the data from when you worked on this, could I please get it from you?
> 
> Many thanks in advance,
> Jules

**Comment by Pelirrojo** [comment]
Hi there,

Thanks for trying my code!
Later in the afternoon I'll try to run it to run and check the error by myself but in advance:

* There is a config file with limits and setup, it's important to keep the limits low to not to be blocked by web server at vgcharts, and doesn't waste/monopolize its bandwidth of course!

* The config file is at: https://github.com/Machine-Learning-Labs/vgchartzScrape/blob/master/cfg/resources.json in this file you can specify the initial and the end page with the register by page. Probably the 200 it's due to the fact that the default config is start_page: 1 and end_page: 2 with 100 register each one. Try to work with these limits.

* The columns are defined as query parameters, but the styles of the website could change, the best way to debug this is to put: 1 page and a few register and print each register, column by column to detect where is the problem (simply printing the value). Also remember that not all the rows has all the information, and some of the information it's represented witch emojis, or images instead text, so needs specific logic to turn into values in final dataset/output.

Hope this to help :)

Thanks,
@Pelirrojo

**Comment by JulesMillen** [comment]
Hi Pelirrojo,

Thank you very much for your prompt and helpful response!

Regarding your second point, I have located what you are referring to and will play around with those limits.

Regarding your third point, I will have a go at this tomorrow (it is already quite late here in NZ).

Regarding your first point, if I was to try and scrape the full website am I likely to be blocked by web server at vgcharts? or is this only an issue if you are scraping the website over and over again?

If it isn't too much trouble, would you mind letting me know how your testing goes?

Many thanks,
Jules

**Comment by Pelirrojo** [comment]
Hi Jules,

I've check the program, and appears to be fine. I've made the next tests:
1) Run directly, just clone and execute: run.sh, with pages 1 to 2, and 100 register each one, with identical results to you: 200 rows.

2) I've modify the config.json file this limits with: pages 1 to 5, and 400 producing 2000 registers.

To ensure the data is right, I've just open with Microsoft Excel and parse in columns (only separating by comma), and the two last columns are always _Release_Date_  and _Last_Update_, normally both fulfilled.
<img width="595" alt="Screenshot 2020-09-08 at 08 42 56" src="https://user-images.githubusercontent.com/2047107/92443291-99334180-f1b1-11ea-871d-7fd0bf5c8438.png">

The main problem it's not all the games has all the data fulfilled, so I'ts normal to found a lot of blanks.

To check the program it's working ok, you can use the uri show in the log when you're running it. This log shows the base uri it's going to use to crawl the website, in my case: 

```log
$>./run.sh
Installing deps...  
Requirement already satisfied: beautifulsoup4==4.8.2 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 1)) (4.8.2) 
Requirement already satisfied: bs4==0.0.1 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 2)) (0.0.1) 
Requirement already satisfied: numpy==1.18.2 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 3)) (1.18.2) 
Requirement already satisfied: pandas==1.0.3 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 4)) (1.0.3) 
Requirement already satisfied: python-dateutil==2.8.1 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 5)) (2.8.1) 
Requirement already satisfied: pytz==2019.3 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 6)) (2019.3) 
Requirement already satisfied: six==1.14.0 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 7)) (1.14.0) 
Requirement already satisfied: soupsieve==2.0 in /Users/pelirrojo/miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 8)) (2.0) 
 
Start crawling... (remember a crawler is the friend nobody likes) 
08-09-2020 08:27:07|root|INFO| Application started 
08-09-2020 08:27:07|root|INFO| generate_remaining_url >>> 
08-09-2020 08:27:07|root|DEBUG| Url Generated: https://www.vgchartz.com/gamedb/?page=N&results=400&region=All&boxart=Both&banner=Both&ownership=Both&showmultiplat=No&order=Sales&showtotalsales=1&showpublisher=1&showvgchartzscore=1&shownasales=1&showdeveloper=1&showcriticscore=1&showpalsales=1&showreleasedate=1&showuserscore=1&showjapansales=1&showlastupdate=1&showothersales=1&showshipped=1&keyword=&console=&developer=&publisher=&goty_year=&genre= 
08-09-2020 08:27:07|root|INFO| generate_remaining_url <<< 
08-09-2020 08:27:07|root|INFO| download_data >>> 
08-09-2020 08:27:07|root|INFO| get_page >>> 
08-09-2020 08:27:07|root|DEBUG| Current URL: https://www.vgchartz.com/gamedb/?page=1&results=400&region=All&boxart=Both&banner=Both&ownership=Both&showmultiplat=No&order=Sales&showtotalsales=1&showpublisher=1&showvgchartzscore=1&shownasales=1&showdeveloper=1&showcriticscore=1&showpalsales=1&showreleasedate=1&showuserscore=1&showjapansales=1&showlastupdate=1&showothersales=1&showshipped=1&keyword=&console=&developer=&publisher=&goty_year=&genre= 
...
```
So the URI base is (take note of the N at page query parameter which will be changing in each page): https://www.vgchartz.com/gamedb/?page=N&results=400&region=All&boxart=Both&banner=Both&ownership=Both&showmultiplat=No&order=Sales&showtotalsales=1&showpublisher=1&showvgchartzscore=1&shownasales=1&showdeveloper=1&showcriticscore=1&showpalsales=1&showreleasedate=1&showuserscore=1&showjapansales=1&showlastupdate=1&showothersales=1&showshipped=1&keyword=&console=&developer=&publisher=&goty_year=&genre= 

And the URI is changing with values included in config file between: _start_page_ and _end_page_ as you supposed, these lines are preceded by _DEBUG: Current URL_. This is useful manually visit web and check everything and in the config file allows to make partial downloads of the web instead downloading in a round.

Finally the file produced is something like this:
<img width="1270" alt="Screenshot 2020-09-08 at 08 43 30" src="https://user-images.githubusercontent.com/2047107/92443321-a51f0380-f1b1-11ea-86be-a41906f25753.png">
and later in the file...
<img width="1266" alt="Screenshot 2020-09-08 at 08 44 57" src="https://user-images.githubusercontent.com/2047107/92443380-b8ca6a00-f1b1-11ea-847e-77cee6730e6e.png">

which correspond with the website with lots of blank values (N/A).
<img width="1351" alt="Screenshot 2020-09-08 at 08 49 01" src="https://user-images.githubusercontent.com/2047107/92443402-c08a0e80-f1b1-11ea-9f71-d31afa29f6c2.png">

I seem to remember that the video game genre is deliberately omitted because it is shown with images rather than text, so it involves adding a specific code to translate that image into a text. It's not complex, but I remember that it often failed and maybe that's why I deleted it.

In any case, I don't remember if the server will blocks you, but it's better to make step by step, because all the registers are recovered from it's database so you're consuming it's resources and if you make in a eager way you can block other visitors of the web.

Hope this help,
Manuel

**Comment by Pelirrojo** [comment]
Another issue I had forgotten, although the configuration file collects all the possible fields of the advanced search form; fields that are then transformed into query parameters of the request URL (i.e. genre=), not all of them worked correctly, so if you activate any of them (setting those that are currently to false to true), you will have to check that they work correctly or make a parser specific for this field, working at this function:

```python
def download_data(*, start_page, end_page, include_genre):
    ...

```

**Comment by JulesMillen** [comment]
Hi Pelirrojo,

Thank you very much for passing that all on, and apologies for a possible false alarm...

That all makes enough sense. I will have a play around with it myself as I need to get Genre too.

Many many thanks!

Jules
